Libplanet/Blockchain/Renderers/DelayedActionRenderer.cs

```diff
@@ -466,6 +506,25 @@ bool unrender
             }
         }
 
+        private void ClearRenderBuffer(
+            long targetBlockIndex,
+            ConcurrentDictionary<HashDigest<SHA256>, List<ActionEvaluation>> buffer)
+        {
+            var invalidityBlock = new List<HashDigest<SHA256>>();
+            foreach (var (hash, evaluations) in buffer.Select(x => (x.Key, x.Value)))
+            {
+                if (evaluations.First().InputContext.BlockIndex < targetBlockIndex)
+                {
+                    invalidityBlock.Add(hash);
+                }
+            }
```

Comment:
I guess LINQ can help us. (you may need to check it again 😅 )

```suggestion
            IEnumerable<HashDigest<SHA256>> invalidityBlockHashes = buffer
                .Select(x => (x.Key, x.Value)))
                .Where(x.Item2.First().InputContext.BlockIndex < targetBlockIndex)
                .Select(t => t.Item1);
```

Reply:
How about...

```suggestion
            IEnumerable<HashDigest<SHA256>> invalidityBlockHashes = buffer
                .Where(x => x.Value.First().InputContext.BlockIndex < targetBlockIndex)
                .Select(t => t.Key);
```

Reply:
chage at a3b8adb